src/module.c

```diff
@@ -6217,18 +6279,31 @@ RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const ch
 
     user *user = NULL;
     if (flags & REDISMODULE_ARGV_RUN_AS_USER) {
-        user = ctx->user ? ctx->user->user : ctx->client->user;
+        user = ctx->client->user;
+        if (ctx->persistent_client) {
+            /* only used for ACL check */
+            user = ctx->persistent_client->client->user;
+        } else if (ctx->user) {
+            user = ctx->user->user;
+        }
+
         if (!user) {
             errno = ENOTSUP;
             if (error_as_call_replies) {
                 sds msg = sdsnew("cannot run as user, no user directly attached to context or context's client");
                 reply = callReplyCreateError(msg, ctx);
             }
+            moduleFreeArgv(argv, argc);
```

Comment:
Is this a leak that exists today? If so I believe we should cover it with tests.

Reply:
I believe so (have another PR for it), I'm unsure how to test it though, I'd have to create a disconnected ctx from the network client (is creating a thread safe context doing that?).